### PR TITLE
Add a new failure handling guide to KerasTuner

### PR DESCRIPTION
**Pull request by haifeng-jin**

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR! Please only include the Python file for the first pass.

**Comment by haifeng-jin** [comment]
Done. removed the generated files. So the generated files will go through another PR afterwards?

File: guides/keras_tuner/failed_trials.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+Description: The basics of fault tolerance configurations in KerasTuner.
+Accelerator: GPU
+"""
+
+"""shell
+pip install keras-tuner -q
+"""
+
+"""
+# Introduction
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:15]: `##` for all sections

File: guides/keras_tuner/failed_trials.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
+"""
+Title: Handling failed trials in KerasTuner
+Authors: Haifeng Jin
+Date created: 2023/02/28
+Last modified: 2023/02/28
+Description: The basics of fault tolerance configurations in KerasTuner.
+Accelerator: GPU
+"""
+
+"""shell
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:10]: Move to a `## Setup` section or similar (after intro), together with imports

File: guides/keras_tuner/failed_trials.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+* We define a search space with 2 hyperparameters for the number of units in the
+  2 dense layers.
+* When their product is larger than 800, we raise a `ValueError` for the model
+  too large.
+
+"""
+
+
+def build_model(hp):
+    model = keras.Sequential()
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:64]: I'd recommend using a list passed to the constructor, that's simpler / shorter

File: guides/keras_tuner/failed_trials.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+    # Define the 2 hyperparameters for the units in dense layers
+    model.add(layers.Dense(units=units_1))
+    model.add(layers.Dense(units=units_2))
+
+    model.add(layers.Dense(units=1))
+    model.compile(loss="mse")
+
+    # Raise an error when the model is too large
+    if units_1 * units_2 > 800:
+        raise ValueError("Model too large!")
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:77]: Print `model.count_params()` in the message

File: guides/keras_tuner/failed_trials.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+    # Define the 2 hyperparameters for the units in dense layers
+    model.add(layers.Dense(units=units_1))
+    model.add(layers.Dense(units=units_2))
+
+    model.add(layers.Dense(units=1))
+    model.compile(loss="mse")
+
+    # Raise an error when the model is too large
+    if units_1 * units_2 > 800:
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:76]: Use `model.count_params()` since that's what matters

File: guides/keras_tuner/failed_trials.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+matter what errors are raised while we may still want to retry for other
+unexpected errors. Is there a way to better handle this situation?
+
+We can raise the `FailedTrialError` to skip the retries. Whenever, this error is
+raised, the trial would not be retried. The retries will still run when other
+errors occur. An example is shown as follows.
+"""
+
+
+def build_model(hp):
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:128]: Same comments for the other `build_model` functions

File: guides/keras_tuner/failed_trials.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    tuner.search(
+        x=np.random.rand(100, 20),
+        y=np.random.rand(100, 1),
+        validation_data=(
+            np.random.rand(100, 20),
+            np.random.rand(100, 1),
+        ),
+        epochs=10,
+    )
+except keras_tuner.errors.FatalError:
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:220]: Please add a short "takeaways" section at the end to recapitulate what the reader has learned.

**Review by fchollet — COMMENTED** [review]
Thanks for the PR! It looks great.

File: guides/keras_tuner/failed_trials.py (hunk 7 above)
Comment (haifeng-jin) [reply guides/keras_tuner/failed_trials.py:240]: Done.

File: guides/keras_tuner/failed_trials.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+        [
+            layers.Dense(units=units_1, input_shape=(20,)),
+            layers.Dense(units=units_2),
+            layers.Dense(units=1),
+        ]
+    )
+    model.compile(loss="mse")
+
+    # Raise an error when the model is too large
+    n_params = model.count_params()
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:152]: Please `num_` as the number prefix, for consistency.

File: guides/keras_tuner/failed_trials.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        ),
+        epochs=10,
+    )
+except keras_tuner.errors.FatalError:
+    print("The search is terminated.")
+
+"""
+## Takeaways
+
+In this tutorial, we learned the following things about handling failed trials
```
Comment (fchollet) [inline guides/keras_tuner/failed_trials.py:246]: "In this guide, you learn how to handle failed trials in KerasTuner:"

**Review by fchollet — COMMENTED** [review]
Thanks for the update! Everything looks great. Two minor comments. Once addressed, please generate the md and ipynb files.

File: guides/keras_tuner/failed_trials.py (hunk 8 above)
Comment (haifeng-jin) [reply guides/keras_tuner/failed_trials.py]: Done.

File: guides/keras_tuner/failed_trials.py (hunk 9 above)
Comment (haifeng-jin) [reply guides/keras_tuner/failed_trials.py]: Done. Why use "learn" instead of "learned" here?

**Review by fchollet — APPROVED** [review]
Thank you! LGTM.

**Pull request merged**